File: algoDat/own/blatt-01/Merge.cpp
// Copyright 2017, University of Freiburg,
// Chair of Algorithms and Data Structures.
// Author: Hannah Bast <bast@cs.uni-freiburg.de>
//         Felix Karg  <felix.karg@uranus.uni-freiburg.de>

#include <iostream>
#include <string>

#include <vector>
#include "./Merge.h"

// ____________________________________________________________________________
// original merge-function written by Bast
std::vector<int> merge2(std::vector<int> v1, std::vector<int> v2) {
    std::size_t n1 = v1.size();
    std::size_t n2 = v2.size();
    std::size_t i1 = 0;
    std::size_t i2 = 0;
    std::size_t k = 0;
    std::vector<int> result(n1 + n2);
    while (i1 < n1 || i2 < n2) {
        if (i1 < n1 && (i2 == n2 || v1[i1] <= v2[i2]))
            result[k++] = v1[i1++];
        if (i2 < n2 && (i1 == n1 || v2[i2] <= v1[i1]))
            result[k++] = v2[i2++];
    }
    return result;
}


// ____________________________________________________________________________
void merge(std::vector<int>& vec, int left, int middle, int right) {
    // wrapper for the other merge-function.
    // completely unnecessary, but o well xD
    // would have been faster with a rewrite though ...

    if (middle > static_cast<int>(vec.size())
            || right > static_cast<int>(vec.size()))
        return;

    std::vector<int> fst(middle - left);
    std::vector<int> snd(right - middle);

    int i = left;

    for (; i < middle; i++) {
        fst[i - left] = vec[i];
    }
    for (; i < right; i++) {
        snd[i - middle] = vec[i];
    }

    // actually sorting the values in the other function
    std::vector<int> sorted = merge2(fst, snd);

    // setting the values to the original vector
    for (int j = left; j < right; j++) {
        vec[j] = sorted[j - left];
    }
}

File: semester_two/algoDat/public/code/vorlesung-01/java/Merge.java
// Copyright 2017, University of Freiburg,
// Chair of Algorithms and Data Structures.
// Author: Hannah Bast <bast@cs.uni-freiburg.de>

/**
 * Class for the merge algorithm from Vorlesung 1b.
 */
public class Merge {

  /**
   * Merge function.
   */
  public static int[] merge(int[] a1, int[] a2) {
    int n1 = a1.length;
    int n2 = a2.length;
    int i1 = 0;
    int i2 = 0;
    int[] result = new int[n1 + n2];
    int k = 0;
    // while (i1 < n1 || i2 < n2)  {
    while (k < n1 + n2) {
      if (i1 < n1 && (i2 == n2 || a1[i1] <= a2[i2])) {
        result[k] = a1[i1];
        i1++;
        k++;
      }
      if (i2 < n2 && (i1 == n1 || a2[i2] < a1[i1])) {
        result[k] = a2[i2];
        i2++;
        k++;
      }
    }
    return result;
  }
}
